### PR TITLE
Feature added games aggregator

### PR DESCRIPTION
**Pull request by JonasPTFL**
Implements added games aggregator according to #2.

Concerning the unit tests, I came up with an alternative approach to the structure used in other aggregators and their tests.
Instead of mocking the Playnite API, I simply created a method in the aggregator that takes all the necessary objects and parameters as arguments, rather than fetching them from the Playnite API itself;
`public IReadOnlyCollection<AddedGame> GetAddedGames(IReadOnlyCollection<Game> allGames, int filterYear)`
With this structure, the method can then be easily tested with custom parameters and without the need to mock any Playnite API calls.
What do you think of this approach? 
Also let me know, what you think of my unit tests. I am not an expert in testing, so im not sure, whether they are designed well :)

### CODE REVIEW

File: YearInReview/Model/Aggregators/AddedGamesAggregator.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+			
+			filteredGames = allGames
+				.Where(g => g?.Added != null && g.Added.Value.Year == filterYear)
+				.Select(g => new AddedGame { AddedDate = g.Added.Value, Game = g })
+				.ToList();
+
+			return filteredGames;
+		}
+		
+		public IReadOnlyCollection<AddedGame> GetAddedGames(int filterYear)
```
Comment (SparrowBrain) [inline YearInReview/Model/Aggregators/AddedGamesAggregator.cs:32]: Good thinking on how not to mock IPlayniteApi :). I would still prefer to have a single method that does everything and where you would not need to provide games from the outside. As in, I would prefer to mock IPlayniteApi.

An example how I inject games into my tests could be found here: `PlaytimeCalendarAggregatorTests.cs:61`

I know that the way I write these tests is quire heavy on various libraries (AutoFixture, FakeItEasy, etc.), so if you want to avoid using AutoFixture with [Frozen] attributes, you could do IPlayniteApi setup in you tests' constructor. In xUnit, constructor is being run independently before each test and tests don't share the context within the test class. So you could do something like:
```
private readonly IGameDatabaseAPI _gameDatabaseApi;
private readonly AddedGamesAggregator _sut;
public AddedGamesAggregatorTest()
{
    var playniteApi = A.Fake<IPlayniteAPI>();
    _gameDatabaseApi = A.Fake<IGameDatabaseAPI>();

    A.CallTo(() => playniteApi.Database).Returns(_gameDatabaseApi);
    _sut = new AddedGamesAggregator(playniteApi);
}
```

And then in the test:
```
    A.CallTo(() => _gameDatabaseApi.Games).Returns(new TestableItemCollection<Game>(games));
```

Now, the reason to avoid using constructors in you code, and instead use AutoFixture is to reduce test fragility. Imagine you want to add a new parameter to you constructor. Now you would have to change all tests where you use that constructor in your setup. Let's rewrite the previous code with AutoFixture:
```
private readonly IGameDatabaseAPI _gameDatabaseApi;
private readonly AddedGamesAggregator _sut;
public AddedGamesAggregatorTest()
{
    var fixture = new Fixture();
    var playniteApi = A.Fake<IPlayniteAPI>();
    _gameDatabaseApi = A.Fake<IGameDatabaseAPI>();

    A.CallTo(() => playniteApi.Database).Returns(_gameDatabaseApi);
    
    // Register an instance of IPlayniteAPI. Now every time we retrieve an instance of IPlayniteAPI it will be our playniteApi variable. So we can configure and use it in our system under test.
    fixture.Register(playniteApi);

    _sut = fixture.Create<AddedGamesAggregator>(); // Here we go. No constructor invocation. Less fragile.
}
```

Honestly, I rarely use AutoFixture like that. As you've seen, my preferred method is to inject variables using autofixture via method parameters. Like:
```
[Theory]
[AutoFakeItEasyData]
public void YourTest_Somethig_WhenSomething(
	[Frozen] IGameDatabaseAPI gameDatabaseApi,
	[Frozen] IPlayniteAPI playniteApi,
	PlaytimeCalendarAggregator sut)
{
...
	A.CallTo(() => gameDatabaseApiFake.Games).Returns(new TestableItemCollection<Game>(games));
	A.CallTo(() => playniteApiFake.Database).Returns(gameDatabaseApiFake);
```
Frozen attribute is basically same as Create() + Register(). Using [Frozen], we first create an instance of IGameDatabaseAPI and then we register it in AutoFixture so that it is used when creating variables during the method call. Also, in this method injection case the fixture is configured, so that for interfaces it will call FakeItEasy and create fakes. That magic resides in `AutoFakeItEasyDataAttribute`.

I understand this might be too much too soon, if you haven't dealt with these .NET libraries (trust me, people who have worked for decades with .NET usually struggle, if they haven't encountered these libraries). But I encourage to try using one of these approaches, so we could avoid injecting games from outside :)

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+			var games = new List<Game>();
+			var game = new Game { Added = date };
+			games.Add(game);
+
+			// Act
+			var result = sut.GetAddedGames(games, date.Year);
+
+			// Assert
+			Assert.Single(result);
+			Assert.True(game.Added == result.First().AddedDate);
```
Comment (SparrowBrain) [inline YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:105]: Prefer `Assert.Equal(expected, actual)`. So
`Assert.Equal(game.Added, single.AddedDate);

If assertion fails you will get more informative message (instead of "is false when expecting true" you will get something like "expected 2025, was 2012").

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+			var date = new DateTime(2025, 1, 1);
+			var games = new List<Game>();
+			var game = new Game { Added = date };
+			games.Add(game);
+
+			// Act
+			var result = sut.GetAddedGames(games, date.Year);
+
+			// Assert
+			Assert.Single(result);
```
Comment (SparrowBrain) [inline YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:104]: `Assert.Single()` returns a value that you can use in your next assertions.

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs (hunk 2 above)
Comment (SparrowBrain) [inline YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:105]: To be fair, if you're checking the game, there is no point to also check the date on the game ;)

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+			Assert.True(dateMatching == result.First().AddedDate);
+			Assert.True(Equals(matchingGame, result.First().Game));
+		}
+
+		[Theory] 
+		[AutoFakeItEasyData] 
+		public void GetAddedGames_ReturnsAllCorrect_WhenAllGamesMatching_AllDaysInYear(AddedGamesAggregator sut)
```
Comment (SparrowBrain) [inline YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:148]: I'm not sure there is a point to check the whole year. Is there a reason you want to do it?

You could as well do 100 or 500 games, seems arbitrary. I would probably remove it :)

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+			// Assert
+			Assert.Equal(3, result.Count);
+			Assert.Contains(result, r => Equals(r.Game, game1) && r.AddedDate == date);
+			Assert.Contains(result, r => Equals(r.Game, game2) && r.AddedDate == date);
+			Assert.Contains(result, r => Equals(r.Game, game3) && r.AddedDate == date);
+		}
+
+		[Theory]
+		[AutoFakeItEasyData]
+		public void GetAddedGames_ReturnsOnlyMatching_WhenIncludesGamesAddedOnBoundaryDates(AddedGamesAggregator sut)
```
Comment (SparrowBrain) [inline YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:206]: This is actually a very clever test. I like it, at the same time, I don't know if it is needed. In reality we will only check the Year on the DateTime :D.

But we can keep it, it's a fun test.

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+			Assert.Equal(2, result.Count);
+			Assert.Contains(result, r => Equals(r.Game, gameOnFirstDay) && r.AddedDate == firstDayOfYear);
+			Assert.Contains(result, r => Equals(r.Game, gameOnLastDay) && r.AddedDate == lastDayOfYear);
+			Assert.DoesNotContain(result, r => Equals(r.Game, gameOutsideYear1));
+			Assert.DoesNotContain(result, r => Equals(r.Game, gameOutsideYear2));
+		}
+		
+		[Theory]
+		[AutoFakeItEasyData]
+		public void GetAddedGames_ReturnsCorrectGames_WhenIncludesDifferentNullValues(AddedGamesAggregator sut)
```
Comment (SparrowBrain) [inline YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:232]: I think either in Roy Osherove's "The Art of Unit Testing" or in another authors "Working with Legacy Code" one of the characteristics of a good unit test was described as: it's precise and helps to pinpoint an issue when it fails.

Since this test is testing a combined rule set, if this test was failing it would not be clear what went wrong. While you might want to have a more "production" like inputs, it could also introduce brittleness (tests break easier, are harder to maintain) and again, it does not pinpoint specific issue.

Food for thought ;)

File: YearInReview/Model/Aggregators/AddedGamesAggregator.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+		public AddedGamesAggregator(IPlayniteAPI playniteApi)
+		{
+			_playniteApi = playniteApi;
+		}
+		
+		public IReadOnlyCollection<AddedGame> GetAddedGames(IReadOnlyCollection<Game> allGames, int filterYear)
+		{
+			var filteredGames = new List<AddedGame>();
+			if (allGames == null || allGames.Count <= 0) return filteredGames;
```
Comment (SparrowBrain) [inline YearInReview/Model/Aggregators/AddedGamesAggregator.cs:22]: I prefer
```
if (allGames == null || allGames.Count <= 0) 
{
    return filteredGames;
}
```

I avoid using inline returns for easier to read code.

**Review by SparrowBrain — COMMENTED** [review]
Pretty good. Left some walls of text as comments. Some style changes, some intro into libraries I use and then some unit test theory I try to follow.

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+			// Act
+			var result = sut.GetAddedGames(games, 0);
+
+			// Assert
+			Assert.Empty(result);
+		}
+
+		[Theory]
+		[AutoFakeItEasyData]
+		public void GetAdded_Empty_WhenYearNegative(AddedGamesAggregator sut)
```
[the remark this answers is not in the document]
Comment (JonasPTFL) [reply YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:62]: Ok, you are right. I will remove this test :)

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+using Playnite.SDK.Models;
+using TestTools.Shared;
+using Xunit;
+using YearInReview.Model.Aggregators;
+
+namespace YearInReview.UnitTests.Model.Aggregators
+{
+	public class AddedGamesAggregatorTest
+	{
+		private static IEnumerable<DateTime> GetAllDatesOfYear(int year)
```
[the remark this answers is not in the document]
Comment (JonasPTFL) [reply YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:13]: Didnt knew that ;D 
I will move it below all public test methods.

File: YearInReview/Model/Aggregators/AddedGamesAggregator.cs (hunk 1 above)
Comment (JonasPTFL) [reply YearInReview/Model/Aggregators/AddedGamesAggregator.cs:32]: Thank you for the explanation about the testing libraries. 
On the one hand side, I personally think its much easier to understand when testing the method directly with the given parameters instead of using mocking. And we are independent of Playnite API, which may change in future. Then we would have to rewrite all tests. 
But on the other hand side, i kind of get your point and you are the experienced C#/.NET developer. And as i dont know much about all this, i will just accept your suggestion :)
Thanks for describing this in detail for me 👍

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs (hunk 2 above)
Comment (JonasPTFL) [reply YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:105]: That makes sense. I will change it.

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs (hunk 3 above)
Comment (JonasPTFL) [reply YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:104]: Thanks, didnt knew that either.

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs (hunk 2 above)
Comment (JonasPTFL) [reply YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:105]: Well yeah, thanks for pointing out. This is stupid ;D

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs (hunk 4 above)
Comment (JonasPTFL) [reply YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:148]: I just wanted to brute force any possible date in one year. I thought, if it is possible to test all possible inputs for a parameter, then check for everything. But on the other hand side, it does not check all values, as it just tests all dates for a single year. So you are right, i can remove it as well.

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs (hunk 5 above)
Comment (JonasPTFL) [reply YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:206]: Yeah you are right. But a test should be designed to test all possible solutions, right? Maybe we change the implemented code to work differently and then it is important to check edge cases. Anyway, I just remembered from my software engineering class at university a year ago how important it is to check for edge cases. So, I thought, why not include something about that here? I will then keep it for now.

File: YearInReview/Model/Aggregators/AddedGamesAggregator.cs (hunk 7 above)
Comment (JonasPTFL) [reply YearInReview/Model/Aggregators/AddedGamesAggregator.cs:22]: Yeah agreed. I will try to keep this in mindwhen writing if statements.

File: YearInReview/Model/Aggregators/AddedGamesAggregator.cs (hunk 1 above)
Comment (SparrowBrain) [reply YearInReview/Model/Aggregators/AddedGamesAggregator.cs:32]: Take note, I just found out I should probably have user `fixture.Inject()` instead of `fixture.Register()`.

**Comment by JonasPTFL** [comment]
I just implemented the changes according to your comments. Let me know if there is still anything you would like to see changed or if I have missed something.

**Comment by JonasPTFL** [comment]
I have synced this branch with your changes from the base repository and resolved the merge conflicts of this pull request.

File: YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+            // Assert
+            Assert.Equal(2, result.Count);
+            Assert.Contains(result, r => Equals(r.Game, gameOnFirstDay) && r.AddedDate == firstDayOfYear);
+            Assert.Contains(result, r => Equals(r.Game, gameOnLastDay) && r.AddedDate == lastDayOfYear);
+            Assert.DoesNotContain(result, r => Equals(r.Game, gameOutsideYear1));
+            Assert.DoesNotContain(result, r => Equals(r.Game, gameOutsideYear2));
+        }
+
+        private static void MockPlayniteApiGames(
```
Comment (SparrowBrain) [inline YearInReview.UnitTests/Model/Aggregators/AddedGamesAggregatorTest.cs:218]: Very good idea!

**Review by SparrowBrain — APPROVED** [review]
